### PR TITLE
add recipe for improved-rdock

### PR DESCRIPTION
**Pull request by otsukay**
Hi, this PR adds a new recipe for [improve-rDock](https://github.com/clinfo/improved_rDock), which is, just as the name suggests, an improved version of [rDock](http://rdock.sourceforge.net/).

### CODE REVIEW

File: var/spack/repos/builtin/packages/improved-rdock/package.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/improved-rdock/package.py]:
```suggestion
# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
```

**Review by tldahlgren — COMMENTED** [review]
There are a lot of patches here. Is there any chance of getting these changes upstreamed?

**Comment by tldahlgren** [comment]
@otsukay Are you aware of https://github.com/CBDD/rDock/?  It is newer than the SourceForge repository though doesn't look active (last post October 2019) either.  Perhaps it would be useful to find out if they'd be willing to upstream the changes?

(Thanks to @wyphan for finding this.  See post at https://spackpm.slack.com/archives/G7XED8NAD/p1655409301102419?thread_ts=1655401974.447389&cid=G7XED8NAD.)

**Comment by otsukay** [comment]
@tldahlgren Thank you for your suggestions.

I confirmed that the developer of improved-rdock does not plan to upstream the patches, mainly because the original code has not been actively maintained, and the improvements cover so many (and rather particular) points that he would like to develop it as an independent project.

(Sorry that I should have explained my position: I am just a maintainer of Spack on Fugaku as a member of @h-murai team at RIKEN , not the developer of the code.)

**Comment by tldahlgren** [comment]
> @tldahlgren Thank you for your suggestions.
> 
> I confirmed that the developer of improved-rdock does not plan to upstream the patches, mainly because the original code has not been actively maintained, and the improvements cover so many (and rather particular) points that he would like to develop it as an independent project.
> 
> (Sorry that I should have explained my position: I am just a maintainer of Spack on Fugaku as a member of @h-murai team at RIKEN , not the developer of the code.)

Thank you for taking the time to check *and* for the clarification.

File: var/spack/repos/builtin/packages/improved-rdock/package.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    """rDock is a fast and versatile Open Source docking program
+    that can be used to dock small molecules against proteins and
+    nucleic acids. """
+
+    homepage = "https://github.com/clinfo/improved_rDock"
+    # url      = "https://sourceforge.net/projects/rdock/files/rDock_2013.1_src.tar.gz"
+    # url = "file://{0}/rDock_2013.1_src_AdvanceSoft_2018.tar.gz".format(os.getcwd())
+    git = "https://github.com/clinfo/improved_rDock.git"
+
+    # version('2013.1', sha256='33eb3aa0c4ede3efe275eb7b7f98c8cb54b0f54d774f400e00cb172e7921b99c')
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/improved-rdock/package.py]: We generally prefer to not retain comments like these.  Perhaps modifying the docstring to include the original repository would be appropriate if you want to retain this information.

File: var/spack/repos/builtin/packages/improved-rdock/package.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        test_dir = self.test_suite.current_test_data_dir
+        copy(join_path(self.prefix.example, '1sj0', '*'), test_dir)
+        opts = ['-r', '1sj0_rdock.prm', '-was']
+        self.run_test('rbcavity', options=opts, work_dir=test_dir)
+
+        mpiexe = self.spec['mpi'].prefix.bin.mpirun
+        opts = [self.prefix.bin.rbdock, '-r', '1sj0_rdock.prm',
+                '-p', 'dock.prm', '-n', '100', '-i', '1sj0_ligand.sd',
+                '-o', '1sj0_docking_out', '-s', '1']
+        self.run_test(mpiexe, options=opts, work_dir=test_dir)
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/improved-rdock/package.py]: The first parameter is supposed to be a string, perhaps:

```suggestion
        self.run_test(str(mpiexe), options=opts, work_dir=test_dir)
```

File: var/spack/repos/builtin/packages/improved-rdock/package.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        install_tree('.', prefix)
+
+    def setup_run_environment(self, env):
+        env.set('RBT_ROOT', self.prefix)
+
+    def test(self):
+        test_dir = self.test_suite.current_test_data_dir
+        copy(join_path(self.prefix.example, '1sj0', '*'), test_dir)
+        opts = ['-r', '1sj0_rdock.prm', '-was']
+        self.run_test('rbcavity', options=opts, work_dir=test_dir)
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/improved-rdock/package.py]: Do you need to be running the tests from the tests *data* directory?  

It is a subdirectory (`data`) of the spack-created test stage and contains a copy of `test/test.sh`.

(The default is for tests to run in the spack-created test stage directory so you could leave the `work_dir` argument off or pass it `'.'`.)

File: var/spack/repos/builtin/packages/improved-rdock/package.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        for shfile in find('bin', '*'):
+            set_executable(shfile)
+        install_tree('.', prefix)
+
+    def setup_run_environment(self, env):
+        env.set('RBT_ROOT', self.prefix)
+
+    def test(self):
+        test_dir = self.test_suite.current_test_data_dir
+        copy(join_path(self.prefix.example, '1sj0', '*'), test_dir)
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/improved-rdock/package.py]: You could copy the file(s) to `'.'`, which is the spack-created, stand-alone test stage directory.

**Comment by tldahlgren** [comment]
Note I am having problems installing package dependencies on my system; otherwise, I would run the tests so I could provide more detailed feedback.

File: var/spack/repos/builtin/packages/improved-rdock/package.py (hunk 4 above)
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/improved-rdock/package.py]: The above assumes the `copy` at line 64 also copies to `'.'`.

File: var/spack/repos/builtin/packages/improved-rdock/package.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        opts = ['-r', '1sj0_rdock.prm', '-was']
+        self.run_test('rbcavity', options=opts)
+
+        mpiexe = self.spec['mpi'].prefix.bin.mpirun
+        opts = [self.prefix.bin.rbdock, '-r', '1sj0_rdock.prm',
+                '-p', 'dock.prm', '-n', '100', '-i', '1sj0_ligand.sd',
+                '-o', '1sj0_docking_out', '-s', '1']
+        self.run_test(str(mpiexe), options=opts)
+
+        opts = ['test.sh']
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/improved-rdock/package.py]: The contents of the `test` subdirectory are automatically copied to a subdirectory of the test stage (namely `self.test_suite.current_test_data_dir`) as described at https://spack.readthedocs.io/en/latest/packaging_guide.html#accessing-package-and-test-related-files, where files under `test` are "Staged Custom Package Files".

```suggestion
        opts = [join_path(self.test_suite.current_test_data_dir, 'test.sh')]
```

Once I addressed my `openbabel` build issues (by disabling the GUI), I was able to successfully test the software with the above fix.

**Review by tldahlgren — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/improved-rdock/package.py (hunk 6 above)
Comment (otsukay) [reply var/spack/repos/builtin/packages/improved-rdock/package.py]: @tldahlgren Thank you for your kind suggestion. Also, it is great to hear that you resolved the installing problem.

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/improved-rdock/package.py]: Not sure I actually solved it but the changes I used to disable the GUI are in https://github.com/spack/spack/pull/31456.

**Review by tldahlgren — APPROVED** [review]

**Pull request merged**